content/300-guides/050-database/820-multi-schema.mdx

```diff
@@ -0,0 +1,107 @@
+
+Multiple database schema support is currently only available with the PostgreSQL and SQL Server connectors.
+
+</Admonition>
+
+This guide explains how to include multiple database schemas in your Prisma schema, and how to query across multiple database schemas with Prisma Client.
+
+</TopBlock>
+
+## What are database schemas?
+
+Many database providers allow you to organise database tables into named groups. This can be done to make the logical structure of the data model easier to understand, or to avoid naming collisions between tables.
+
+In PostgreSQL and SQL Server these groups are known as schemas. We will refer to them as _database schemas_ to distinguish them from Prisma's own schema file.
+
+## How to enable the <inlinecode>multiSchema</inlinecode> preview feature
+
+Multi-schema support is currently in an early, experimental preview. To enable the `multiSchema` preview feature, you will need to add the `multiSchema` feature flag to the `previewFeatures` field of the `generator` block in your Prisma schema file:
+
+```prisma file=schema.prisma highlight=3;add
+generator client {
+  provider        = "prisma-client-js"
+  previewFeatures = ["multiSchema"]
+}
+
+datasource db {
+  provider = "postgresql"
+  url      = env("DATABASE_URL")
+}
+```
+
+## How to include multiple database schemas in your Prisma schema
+
+To use multiple database schemas in your Prisma schema file, add the names of your database schemas to an array in the `schemas` field, in the `datasource` block. The following example adds a `"base"` and a `"transactional"` schema:
+
+```prisma file=schema.prisma highlight=9;add
+generator client {
+  provider        = "prisma-client-js"
+  previewFeatures = ["multiSchema"]
+}
+
+datasource db {
+  provider = "postgresql"
+  url      = env("DATABASE_URL")
+  schemas  = ["base", "transactional"]
+}
+```
+
+You do not need to change your connection string. The `schema` value of your connection string is the default database schema that the Prisma Client connects to and uses for raw queries. All other Prisma Client queries will use the schema of the model that you are querying.
+
+To designate a model as belonging to a specific database schema, add a `@@schema` attribute with the name of the database schema as a parameter. In the following example the `User` model is part of the `"base"` schema, and the `Order` model is part of the `"transactional"` schema:
+
+```prisma file=schema.prisma highlight=5,13;add
+model User {
+  id     Int     @id
+  orders Order[]
+
+  @@schema("base")
+}
+
+model Order {
+  id      Int  @id
+  user    User @relation(fields: [id], references: [id])
+  user_id Int
+
+  @@schema("transactional")
+}
+```
+
+<Admonition type="warning">
+
+Currently it is not allowed to have models with the same name that use the `@@schema` attribute to reference different database schemas. All models in your Prisma schema must have unique names. Otherwise, your Prisma schema will fail to validate.
```

Comment:
It makes things a bit more complicated, but I think we should remind users at this point, that models, like most things in a Prisma schema, have two names. For example:

```prisma
model TestA {
  id Int @id
}

model TestB {
  id Int @id
  @@map("test_b")
}
```

`TestA` has:

- client name: `TestA`
- database name: `TestA`

`TestB` has:

- client name: `TestB`
- database name `test_b`

So far, before multi-schema, there is one namespace for client names, and another for database names.

Since multi-schema is about multiple namespaces in the database, model names clash differently.

- Client names still follow the same rules. You can only have one model with a given name in a Prisma schema.
- Database names live inside a schema, so this _should_ be valid:

```prisma
model TestA {
  id Int @id

  @@map("test")
  @@schema("firstSchema")
}

model TestB {
  id Int @id
  @@map("test")
  @@schema("secondSchema")
}
```

Unfortunately, the implementation is incomplete so we still reject this schema, but it will be fixed when we switch back to multi-schema work.

Reply:
@tomhoule thanks for the comment on model names, that's clearer now. Though I'm not sure how to describe it in docs, particularly at the current stage where the example you give isn't implemented yet.

For now I've changed 'the same name' to 'the same name in your Prisma schema' in the admonition to make it a little clearer. It's not exactly the distinction you're pointing at as both the 'client name' and 'database name' can be described with the Prisma schema, but I think most people would interpret what I've written as meaning the client one? Once the example works I'll update the docs and write something more comprehensive. Does that work for you?

Reply:
yes, that sounds good :+1: